pressurecooker/images.py

```diff
@@ -21,50 +25,108 @@
+
+from .thumbscropping import scale_and_crop
 
-    images = list(map(Image.open, source_images))
-    widths, heights = zip(*(i.size for i in images))
-
-    max_dimension = int(min(max(widths), max(heights)))
-    offset = int(max_dimension / int(root))
-
-    new_im = Image.new('RGB', (max_dimension, max_dimension))
-
-    index = x_index = y_index = 0
-    root = int(root)
-    while y_index < root and index < len(images):
-        x_index = 0
-        while x_index < root and index < len(images):
-            im = ImageOps.fit(images[index], (int(offset), int(offset)), Image.ANTIALIAS)
-            new_im.paste(im, (int(offset * x_index), int(offset * y_index)))
-            x_index += 1
-            index += 1
-        y_index += 1
-    new_im.save(fpath_out)
 
-def create_image_from_pdf_page(fpath_in, fpath_out, page_number=0):
+
+# SMARTCROP UTILS
+################################################################################
+
+THUMBNAIL_SIZE = (400, 225)  # 16:9
+
+def scale_and_crop_thumbnail(image, size=THUMBNAIL_SIZE, crop="smart", **kwargs):
+    """
+    Scale and crop the PIL Image ``image`` to maximum dimensions of ``size``.
+    By default, ``crop`` is set to "smart" which will crop the image down to size
+    based on the entropy content of the pixels. The other options are:
+    * Use ``crop="0,0"`` to crop from the left and top edges
+    * Use ``crop=",0"`` to crop from the top edge.
+    Optional keyword arguments:
+    * ``zoom=X``: crop outer X% before starting
+    * ``target``: recenter here before cropping (default center ``(50, 50)``)
+    See the ``scale_and_crop`` docs in ``thumbscropping.py`` for more details.
+    """
+    return scale_and_crop(image, size, crop=crop, upscale=True, **kwargs)
+
+
+
+# THUMBNAILS FOR CONTENT KINDS
+################################################################################
+
+def get_image_from_epub(epubfile, fpath_out, crop=None):
+    book = ebooklib.epub.read_epub(epubfile)
+    # 1. try to get cover image from book metadata (content.opf)
+    cover_item = None
+    covers = book.get_metadata('http://www.idpf.org/2007/opf', 'cover')
+    if covers:
+        cover_tuple = covers[0] # ~= (None, {'name':'cover', 'content':'item1'})
+        assert cover_tuple[1]['name'] == 'cover', 'wrong key name'
+        cover_item_id = cover_tuple[1]['content']
+        for item in book.items:
+            if item.id == cover_item_id:
+                cover_item = item
+    if cover_item:
+        image_data = BytesIO(cover_item.get_content())
+    else:
+        # 2. fallback to get first image in the ePub file
+        images = list(book.get_items_of_type(ebooklib.ITEM_IMAGE))
+        # TODO: get largest image of the bunch
+        if not images:
+            return
+        image_data = BytesIO(images[0].get_content())
+
+    # Save image_data to fpath_out
+    im = Image.open(image_data)
+    im = scale_and_crop_thumbnail(im, crop=crop)
+    im.save(fpath_out)
```

Comment:
@bjester All the thumbnail generating functions have the signature `get_image_from_???(fpath_in, fpath_out) --> None` where two file paths are passed in. The calling code is responsible for managing the lifetime of `fpath_in`  and `fpath_out` (e.g. deleting the temp files as needed).

I'm wondering what's the best way to handle errors in this case, and I'm hestitating between two options:
  1. make conversion function return `success` (bool) that the calling code can check and know not to expect a thumbnail file at `fpath_out` 
  2. add a separate `check_thumbnail_file` helper method that calling code can call to check if `fpath_out`is really a well formed thumbnail.

What do you think would be better/cleaner?

Reply:
What kind of errors are you trying to handle?

Reply:
It's content-type dependent; here are some examples:
  - `.zip` file that contains no images
  - `.epub` file with no cover image and an no image items
  - for PDFs and videos it's less likely to happen (assuming source is a valid file)

In all of the above cases the calling code needs to be notified "Hey, I know you asked me to put a thumnail in `fpath_out` but I wasnt able to do so, so please proceed without a thumnnail" rather than uploading the empty file as thumbnail.

Reply:
Okay, I see! I think option 1, returning a bool, makes sense to me. Did you have concerns with that?

Reply:
Well, on second glance, I guess the function name starting with `get_` and returning a bool could be confusing. If it was `generate_` then I wouldn't give it a second thought, but with `get_` I feel I should have something related specifically to the image.

Reply:
no, i was just hesitating between the two options...

ok will update the PR like that --- it will require only minor changes to calling code to check the `success`  return value.

Reply:
Okay so in the end I decided to redo as Exceptions --- there is a precedent about this for the video compression utility so I'm going to follow this example:
https://github.com/learningequality/pressurecooker/blob/master/pressurecooker/videos.py#L43-L79
(otherwise the calling code was going to get really weird with a mix of boolean status checks + exception handling)

I rename the `get_`... functions to `create_`. Much better naming this way. Thx for flagging that.

Reply:
Okay that works too. I could still see it being okay returning a bool (False) if the functions have nothing to process, but not much help agonizing over it.